### PR TITLE
Enable pull request test filtering

### PR DESCRIPTION
**Pull request by matt-kwong**
Creating pull request to test using Jenkins. This will be ready to merge after being satisfied with the dry run testing and testing how this interacts with Jenkins.

### CODE REVIEW

**Comment by jtattermusch** [comment]
looking at some test results.

We are changing files that should require re-running all the tests, but we skipped everything but node and python.

We should look into the skipping algorithm again.
Btw, optional: perhaps it is a good idea to set up a bunch of tests for the skipping algorithm and put them in the sanity suite?

```
Finding file differences between gRPC origin/master branch and pull request...

  tools/run_tests/filter_pull_request_tests.py
  tools/run_tests/jobset.py
  tools/run_tests/report_utils.py
  tools/run_tests/run_tests_matrix.py

  sanity tests safe to skip
  c tests safe to skip
  c++ tests safe to skip
  csharp tests safe to skip
  objc tests safe to skip
  php tests safe to skip
  php7 tests safe to skip
  ruby tests safe to skip
  linux tests safe to skip
  windows tests safe to skip
  macos tests safe to skip

These tests will be skipped:
  run_tests_csharp_linux_dbg
  run_tests_ruby_linux_dbg
  run_tests_php7_linux_opt
  run_tests_sanity_linux_opt
  run_tests_c_linux_dbg
  run_tests_php7_linux_dbg
  run_tests_php_linux_dbg
  run_tests_c++_linux_opt
  run_tests_csharp_linux_opt
  run_tests_php_linux_opt
  run_tests_c_linux_opt
  run_tests_sanity_linux_dbg
  run_tests_c++_linux_dbg
  run_tests_ruby_linux_opt

Will run these tests:
  run_tests_node_linux_dbg
  run_tests_node_linux_opt
  run_tests_python_linux_dbg
  run_tests_python_linux_opt
```

**Comment by matt-kwong** [comment]
The reason for so many skipped tests is because of this: 
```
 # todo(mattkwong): Remove this
 # Faking changed files to test test filtering on Jenkins
 changed_files = ['src/node/something', 'src/python/something']
```
Without this, all tests would run, and I need some tests to skip to see how it shows up in Jenkins.

I agree about making tests for the skipping algorithm. Related to adding them to sanity tests, I have a feeling sanity tests are being skipped too much, but we should go over the whitelist a few more times during the testing period to make sure we're not skipping important tests.

File: tools/run_tests/filter_pull_request_tests.py [hunk 1]
```diff
@@ -174,9 +175,13 @@ def filter_tests(tests, base_branch):
   print("Finding file differences between gRPC %s branch and pull request...\n" % base_branch)
   changed_files = _get_changed_files(base_branch)
   for changed_file in changed_files:
-    print(changed_file)
+    print("  %s" % changed_file)
   print
 
+  # todo(mattkwong): Remove this
```
Comment (jtattermusch) [inline tools/run_tests/filter_pull_request_tests.py]: nit: let's just remove this entirely

File: tools/run_tests/run_tests_matrix.py [hunk 2]
```diff
@@ -209,100 +209,108 @@ def _allowed_labels():
...
-print 'Will run these tests:'
-for job in jobs:
-  if args.dry_run:
-    print '  %s: "%s"' % (job.shortname, ' '.join(job.cmdline))
-  else:
-    print '  %s' % job.shortname
-print
+  skipped_jobs = []
+  if args.filter_pr_tests:
+    print 'Looking for irrelevant tests to skip...'
```
Comment (jtattermusch) [inline tools/run_tests/run_tests_matrix.py]: nit: better to use print with () brackets as it is python3 compatible. `print('my message')`

File: tools/run_tests/run_tests_matrix.py [hunk 3]
```diff
@@ -209,100 +209,108 @@ def _allowed_labels():
...
 
-if args.dry_run:
-  print '--dry_run was used, exiting'
-  sys.exit(1)
+  if args.dry_run:
+    print '--dry_run was used, exiting'
+    sys.exit(1)
+
+  jobset.message('START', 'Running test matrix.', do_newline=True)
+  num_failures, resultset = jobset.run(jobs,
```
Comment (jtattermusch) [inline tools/run_tests/run_tests_matrix.py]: if there are no jobs to run, what's going to happen?  (likely to happen for many PRs with sanity linux test suites). 
It shouldn't fail and it need to generate report.xml because jenkins will be looking for it (and fails the job if it's not present).

File: tools/run_tests/run_tests_matrix.py [hunk 4]
```diff
@@ -209,100 +209,108 @@ def _allowed_labels():
...
-                                     suite_name='aggregate_tests')
 
-if num_failures == 0:
-  jobset.message('SUCCESS', 'All run_tests.py instance finished successfully.',
-                 do_newline=True)
-else:
-  jobset.message('FAILED', 'Some run_tests.py instance have failed.',
-                 do_newline=True)
-  sys.exit(1)
+  if num_failures == 0:
```
Comment (jtattermusch) [inline tools/run_tests/run_tests_matrix.py]: did you really want this extra level of indentation?

Comment (matt-kwong) [inline tools/run_tests/run_tests_matrix.py]: When I import run_tests_matrix.py into my test .py file, run_tests_matrix.py runs unless the code is put inside 'if __name__ == "__main__":'. I don't think there is any other way to avoid this, and it should be ok, since every other current interaction run_tests_matrix.py calls it directly.

File: tools/run_tests/filter_pull_request_tests.py [hunk 5]
```diff
@@ -174,8 +175,8 @@ def filter_tests(tests, base_branch):
   print("Finding file differences between gRPC %s branch and pull request...\n" % base_branch)
   changed_files = _get_changed_files(base_branch)
   for changed_file in changed_files:
-    print(changed_file)
-  print
+    print("  %s" % changed_file)
```
Comment (jtattermusch) [inline tools/run_tests/filter_pull_request_tests.py]: nit: use consistent quote type. Our other python scripts use mostly `'`, so let's get rid of the double quotes.

Comment (jtattermusch) [inline tools/run_tests/filter_pull_request_tests.py]: Fine to address later.

**Comment by nicolasnoble** [comment]
@ctiller any opinion on the whitelist ?

File: tools/run_tests/filter_pull_request_tests.py [hunk 6]
```diff
@@ -77,6 +77,7 @@ def add_trigger(self, trigger):
 # and the value is a list of tests that should be run. An empty list means that
 # the changed files should not trigger any tests. Any changed file that does not
 # match any of these regexes will trigger all tests
+# DO NOT CHANGE THIS UNLESS YOU KNOW WHAT YOU ARE DOING (be careful even if you do)
 _WHITELIST_DICT = {
```
Comment (nicolasnoble) [inline tools/run_tests/filter_pull_request_tests.py]: @ctiller this whitelist, basically.

**Comment by matt-kwong** [comment]
ping @ctiller 

Any feedback on the whitelist would be great. With this, PR testing should be much faster/smoother, but of course we should be concerned that we are not accidently skipping relevant tests.

**Comment by matt-kwong** [comment]
_WHITELIST_DICT = {
  '^doc/': [],
  '^examples/': [],
  '^include/grpc\+\+/': [_CPP_TEST_SUITE],
  '^summerofcode/': [],
  '^src/cpp/': [_CPP_TEST_SUITE],
  '^src/csharp/': [_CSHARP_TEST_SUITE],
  '^src/node/': [_NODE_TEST_SUITE],
  '^src/objective\-c/': [_OBJC_TEST_SUITE],
  '^src/php/': [_PHP_TEST_SUITE],
  '^src/python/': [_PYTHON_TEST_SUITE],
  '^src/ruby/': [_RUBY_TEST_SUITE],
  '^templates/': [_SANITY_TEST_SUITE],
  '^test/core/': [_CORE_TEST_SUITE],
  '^test/cpp/': [_CPP_TEST_SUITE],
  '^test/distrib/cpp/': [_CPP_TEST_SUITE],
  '^test/distrib/csharp/': [_CSHARP_TEST_SUITE],
  '^test/distrib/node/': [_NODE_TEST_SUITE],
  '^test/distrib/php/': [_PHP_TEST_SUITE],
  '^test/distrib/python/': [_PYTHON_TEST_SUITE],
  '^test/distrib/ruby/': [_RUBY_TEST_SUITE],
  '^vsprojects/': [_WINDOWS_TEST_SUITE],
  'binding\.gyp$': [_NODE_TEST_SUITE],
  'composer\.json$': [_PHP_TEST_SUITE],
  'config\.m4$': [_PHP_TEST_SUITE],
  'CONTRIBUTING\.md$': [],
  'Gemfile$': [_RUBY_TEST_SUITE],
  'grpc\.def$': [_WINDOWS_TEST_SUITE],
  'grpc\.gemspec$': [_RUBY_TEST_SUITE],
  'gRPC\.podspec$': [_OBJC_TEST_SUITE],
  'gRPC\-Core\.podspec$': [_OBJC_TEST_SUITE],
  'gRPC\-ProtoRPC\.podspec$': [_OBJC_TEST_SUITE],
  'gRPC\-RxLibrary\.podspec$': [_OBJC_TEST_SUITE],
  'INSTALL\.md$': [],
  'LICENSE$': [],
  'MANIFEST\.md$': [],
  'package\.json$': [_PHP_TEST_SUITE],
  'package\.xml$': [_PHP_TEST_SUITE],
  'PATENTS$': [],
  'PYTHON\-MANIFEST\.in$': [_PYTHON_TEST_SUITE],
  'README\.md$': [],
  'requirements\.txt$': [_PYTHON_TEST_SUITE],
  'setup\.cfg$': [_PYTHON_TEST_SUITE],
  'setup\.py$': [_PYTHON_TEST_SUITE]
}

Basically this thing. If any changed files does not match one of these regexes, then all tests will be run. Otherwise, the only tests run are the ones inside the value of the keys matched in the dict. If a value of a key is empty, then that file being changed should not trigger any tests.

File: tools/run_tests/filter_pull_request_tests.py [hunk 7]
```diff
@@ -104,7 +105,7 @@ def add_trigger(self, trigger):
   'config\.m4$': [_PHP_TEST_SUITE],
   'CONTRIBUTING\.md$': [],
   'Gemfile$': [_RUBY_TEST_SUITE],
-  'grpc.def$': [_WINDOWS_TEST_SUITE],
+  'grpc\.def$': [_WINDOWS_TEST_SUITE],
   'grpc\.gemspec$': [_RUBY_TEST_SUITE],
```
Comment (jtattermusch) [inline tools/run_tests/filter_pull_request_tests.py]: One case that these definitions are missing is when someone manually updates files that are supposed to be autogenerated.  (e.g.  files under /vsprojects).  In that case, sanity tests won't run, yet from my experience people often forget that some files are generated and update them manually - and we want to prevent that from happening.

You probably need to add sanity test run for all updates to generated files (which is somewhat tricky, because there is a lot of them and more can be added later - without having a way to enumerate them, we might need to just add _SANITY_TEST_SUITE to all whitelist entries)
E.g.
`'^vsprojects/': [_WINDOWS_TEST_SUITE, _SANITY_TEST_SUITE],`

Actually, I don't understand why sanity takes such long time to run - I think we should be running it as often as possible.

Comment (matt-kwong) [inline tools/run_tests/filter_pull_request_tests.py]: In that case, would it be best to just remove skipping sanity tests completely to be safe? RIght now, the bottleneck for pull request tests are interop and performance, so we could worry about sanity tests taking too long when the basic tests become the bottleneck.

**Comment by matt-kwong** [comment]
test this please
